### DISCUSSION TITLE
XYRangeIndicator

### DISCUSSION THREAD

**Discussion by Renegadezz** [GSI-CS-CO/chart-fx#390] [Ideas]
Hi loving the library.
Tho I thought some functionality was missing such as an XYRangeIndicator.
You currently have both the X - & Y-RangeIndicators plugin but not one for both axes at the same time, which seems like something most people could use as well. 
I've gone ahead and made a MWE tho keep in mind it's made in Kotlin

```Kotlin
class XYRangeIndicator(
    xAxis: Axis,
    var xLowerBound: Double,
    var xUpperBound: Double,
    var yLowerBound: Double,
    var yUpperBound: Double,
    text: String? = null
) : AbstractRangeValueIndicator(xAxis, xLowerBound, xUpperBound, text) {

    override fun layoutChildren() {
        if (chart == null) {
            return
        }
        val plotAreaBounds = chart.canvas.boundsInLocal
        val minX = plotAreaBounds.minX
        val maxX = plotAreaBounds.maxX
        val minY = plotAreaBounds.minY
        val maxY = plotAreaBounds.maxY

        val xAxis = chart.getFirstAxis(Orientation.HORIZONTAL)
        val xLBPos = xAxis.getDisplayPosition(xLowerBound)
        val xUBPos = xAxis.getDisplayPosition(xUpperBound)

        val yAxis = chart.getFirstAxis(Orientation.VERTICAL)
        val yLBPos = yAxis.getDisplayPosition(yLowerBound)
        val yUBPos = yAxis.getDisplayPosition(yUpperBound)

        val startX = Math.max(minX, minX + Math.min(xLBPos, xUBPos))
        val endX = Math.min(maxX, minX + Math.max(xLBPos, xUBPos))

        val startY = Math.max(minY, minY + Math.min(yLBPos, yUBPos))
        val endY = Math.min(maxY, minY + Math.max(yLBPos, yUBPos))

        layout(BoundingBox(startX, startY, endX - startX, endY - startY))
    }

    override fun updateStyleClass() {
        setStyleClasses(label, "x-", STYLE_CLASS_LABEL)
        setStyleClasses(rectangle, "x-", STYLE_CLASS_RECT)
    }
}
```

This is more of a workaround than a final solution as I instantiate an AbstractRangeValueIndicator with only the Xaxis and X-lower and upper bounds, tho the Yaxis and Y-lower and upper bounds are used to draw the square. 
This is made from the JDK8 branch.

**Comment by RalphSteinhagen** [comment]
@Blazingsteal thanks for your interest. 

If I understand your use-case correctly, you want to mark a given area inside the plot using a bounding box -- similar to what the zoomer already does -- but without performing the actual zoom at the end and while providing properties for the corners in units of the given axes. 

I.e. something like
```Java
 List<double> getBoundingBoxMinX();  // list of coordinates for the 1st, 2nd, ... nth horizontal axis
 List<double> getBoundingBoxMinY(); // list of coordinates for the 1st, 2nd, ... nth vertical axis
 List<double> getBoundingBoxMaxX();
 List<double> getBoundingBoxMaxY();
```
If yes, then the mouse- and bounding box handling of the [Zoomer](https://github.com/GSI-CS-CO/chart-fx/blob/master/chartfx-chart/src/main/java/de/gsi/chart/plugins/Zoomer.java) plugin notably the [`Rectangle zoomRectangle`](https://github.com/GSI-CS-CO/chart-fx/blob/8fd26671ecd797718baa90fc6cb660e8489bd993/chartfx-chart/src/main/java/de/gsi/chart/plugins/Zoomer.java#L145) might perhaps be a better starting point? This would provide also an (optional) history of the bound box. 

What would you do with the corners of the bounding box? Does your use-case require other UI elements (e.g. the XValueIndicators have additional markers on the axes). Do you need this value to derive/change other objects within or outside the chart? 

Since the above API doesn't exist (yet), this could be used to shape and influence a bit the final API design choices such as lists of doubles, tuples, domain objects containing also information on the axes like unit, etc.

Would you be interested in this and willing to make a PR? Let us know.

**Reply by RalphSteinhagen** [reply]
May I ask what type of project/domain you intend to use this? This would help .... thanks in advance.

**Reply by Renegadezz** [reply]
> If I understand your use-case correctly, you want to mark a given area inside the plot using a bounding box -- similar to what the zoomer already does -- but without performing the actual zoom at the end and while providing properties for the corners in units of the given axes.

Yes. Ideally what the feature would be is: you being able to mark a square on the chart, either through marking with the zoomer (as you mentioned) or giving lower and upper boundaries for the axes, and applying a label to that selection. 

> If yes, then the mouse- and bounding box handling of the Zoomer plugin notably the Rectangle zoomRectangle might perhaps be a better starting point? This would provide also an (optional) history of the bound box.

Just for marking an area, definitely. For adding label an actual XY-Indicator might be better.

> What would you do with the corners of the bounding box? Does your use-case require other UI elements (e.g. the XValueIndicators have additional markers on the axes). Do you need this value to derive/change other objects within or outside the chart?

The markers on the axes are not something I've used for anything, nor do I require any other UI elements besides the label that can be applied to the indicators.

Currently I'm using your library for a project where I handle data (sometimes live data) of vegetables. This includes but not limited to, length/diameter/volume etc. which then is shown in several graphs/charts.

And I'm using the XYIndicator to show product count, averages etc. in the given range in a chart where e.g. X-axis is length and Y-axis is diameter.

As for the PR I'm not sure as I currently develop most my stuff in kotlin (though very compatible with and similar to java) nor can I do it on company time. I only had time to make this because it was a request for the project.

**Reply by RalphSteinhagen** [reply]
Do you have a screenshot of your workaround in action?

**Reply by Renegadezz** [reply]
[Yes](https://imgur.com/a/QDRYTUs)
This is without any text from the user, so there will always be a label.
